### PR TITLE
[SR] Support Jetpack Compose redaction

### PR DESCRIPTION
**Pull request by romtsn**
## :scroll: Description

* Add support for jetpack compose redaction
* Unfortunately we had to use quite some reflection and internal kotlin classes as the means that compose exposes to public are not enough. Namely, I've tried to use the semantics tree [approach](https://blog.bitdrift.io/post/implementing-session-replay-android-compose), however if a compose node does not have any accessibility property set, it's missing entirely from the semantics tree. So instead we:
  * Traverse the real tree (LayoutNodes)
  * Use the corresponding semantics node (if exists), to fetch some valuable information like TextLayout to later use it for masking
  * When it's not possible to reliably retrieve the info from the semantics node (e.g. Images), we fall back to reflection to fetch the properties we need (e.g. Image's `painter`), as they are marked as private
  * The reflection calls should be fairly lightweight, as we actually only checking for classnames in the for-loop, but then call `getDeclaredField` only once per Image/Text in the tree. I've also tested this in a release version of pocketcasts and tree parsing was well under 16ms, which makes it actually faster than the old view system traversal at times. We can think of using bytecode manipulation later to make it even faster.
  * The necessary proguard rules were also added and tested in a release build of pocketcasts
* Also added custom modifiers `sentryReplayRedact`/`sentryReplayIgnore` to allow selective masking 
* We only support compose 1.5.0 or above, as the entire semantics logic was reworked in this version and it'd be hard to support the prior ones. Should be fine since the version is > 1 year old.
* We added a `compileOnly` dependency on compose-ui (the only one we need). Should be fine since we check the AndroidComposeView presence at runtime and skipping the logic altogether if it's not available.

### Link to sample replay
https://sentry-sdks.sentry.io/replays/a5ebb46afb5c41cd82a7ebb6b1762943/?project=5428559&query=&referrer=%2Freplays%2F%3AreplaySlug%2F&statsPeriod=30d&yAxis=count%28%29&t=0

### Example screenshot of a fully-compose screen

image

## :bulb: Motivation and Context

Part of https://github.com/getsentry/sentry/issues/74441
Closes https://github.com/getsentry/sentry-java/issues/3577

## :green_heart: How did you test it?
Manually + automated

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,18 +6,20 @@
 
 - Add support for `feedback` envelope header item type ([#3687](https://github.com/getsentry/sentry-java/pull/3687))
 - Add breadcrumb.origin field ([#3727](https://github.com/getsentry/sentry-java/pull/3727))
+- Session Replay: Add options to selectively redact/ignore views from being captured. The following options are available: ([#3689](https://github.com/getsentry/sentry-java/pull/3689))
```
Comment (romtsn) [inline CHANGELOG.md:9]: moved this as it was wrongly added to the Fixes section

File: sentry-android-replay/src/main/java/io/sentry/android/replay/ModifierExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package io.sentry.android.replay
+
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.semantics.SemanticsPropertyKey
+import androidx.compose.ui.semantics.semantics
+import io.sentry.android.replay.SentryReplayModifiers.SentryPrivacy
+
+public object SentryReplayModifiers {
```
Comment (markushi) [inline sentry-android-replay/src/main/java/io/sentry/android/replay/ModifierExtensions.kt:8]: Nice!

File: sentry-android-replay/src/main/java/io/sentry/android/replay/util/Nodes.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    val modifierInfos = getModifierInfo()
+    var color: Color? = null
+    var hasFillModifier = false
+    for (index in modifierInfos.indices) {
+        val modifier = modifierInfos[index].modifier
+        val modifierClassName = modifier::class.java.name
+        if (modifierClassName.contains("Text")) {
+            color = try {
+                (
+                    modifier::class.java.getDeclaredField("color")
```
Comment (markushi) [inline sentry-android-replay/src/main/java/io/sentry/android/replay/util/Nodes.kt:114]: maybe it makes sense to cache the field outside of `findTextAttributes()`

File: sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (markushi) [inline sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml:1]: Should we add an attribution header here? Something along those lines?
```suggestion
<!-- Based on https://github.com/Automattic/pocket-casts-android/blob/main/modules/services/images/src/main/res/drawable/logo_pocket_casts.xml 

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at [http://mozilla.org/MPL/2.0/.](http://mozilla.org/MPL/2.0/)
-->
<vector xmlns:android="http://schemas.android.com/apk/res/android"
```

**Review by markushi — APPROVED** [review]
Hats off, nicely done!

File: sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml (hunk 4 above)
Comment (romtsn) [reply sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml:1]: ow yes, a very good callout!

File: sentry-android-replay/src/main/java/io/sentry/android/replay/util/Nodes.kt (hunk 3 above)
Comment (romtsn) [reply sentry-android-replay/src/main/java/io/sentry/android/replay/util/Nodes.kt:114]: yep, I thought about the same, but the problem here is that the holding class is actually private, so we'd have to access it via reflection too (`Class.forName`). And in this case we actually don't know what the class is until we start iterating over the modifiers list (there can be multiple classes holding this field), so we'd have to cache the field for each possible class, and the gains are questionable in this case.

in my tests it was quite fast (like 0.05ms on average for this method to execute), which resulted in the entire tree being parse within 4.5ms for an average screen. I think I'd rather focus on making it work via bytecode manipulation, so we wipe this entirely, wdyt?

```
E_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.663  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.664  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.664  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.665  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.666  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.12 ms
2024-10-08 17:53:47.667  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.667  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.671  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.19 ms
2024-10-08 17:53:47.672  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.673  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.22 ms
2024-10-08 17:53:47.674  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.674  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.675  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.675  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.676  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.678  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.679  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.680  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.680  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.681  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.681  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.683  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.685  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.16 ms
2024-10-08 17:53:47.685  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.686  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.686  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.687  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.689  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.690  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.690  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.693  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.694  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.694  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.695  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.697  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.699  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.700  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.700  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.701  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.701  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.702  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.704  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.704  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.705  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.705  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.706  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.706  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.708  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.709  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.12 ms
2024-10-08 17:53:47.710  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.714  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.715  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.715  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.719  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.720  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.720  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.724  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.724  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.725  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.725  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.726  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.726  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.728  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.729  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.729  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.732  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.733  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.733  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.00 ms
2024-10-08 17:53:47.736  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
```

File: sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml (hunk 4 above)
Comment (romtsn) [reply sentry-samples/sentry-samples-android/src/main/res/drawable/logo_pocket_casts.xml:1]: I actually realized that you cannot use the contributor's logo according to MPL, so I'll just replace it with the sentry one 😅

**Pull request merged**